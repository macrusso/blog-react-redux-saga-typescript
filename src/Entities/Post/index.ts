import * as postConstants from "./postConstants";
import * as postSelectors from "./postSelectors";
import * as postActions from "./postActions";
import * as postSagas from "./postSagas";
import postReducer from "./postReducer";
import Post from "../../Frontend/Post/PostListItem";
import IPostPartial from "./IPostPartial";
import IPost from "./IPost";
import IPostsState from "./IPostsState";

export { Post, IPost, postSagas, postActions, IPostsState, postReducer, IPostPartial, postSelectors, postConstants };
