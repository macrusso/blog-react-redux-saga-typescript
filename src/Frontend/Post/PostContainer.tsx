import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import { IAction, IStoreState } from '../../types';
import ErrorBoundary from '../Common/ErrorBoundary';
import {
  userActions,
  userSelectors,
  IUser,
  postSelectors,
  IPost,
  postActions,
} from '../../Entities';
import {
  CommentListContainer,
  CommentAddContainer,
  PostEditDialog,
  Post,
  PostDeleteDialog,
} from '../../Frontend';
import { Route } from 'react-router';
import * as routes from '../../routes';
import { push } from 'connected-react-router';

type IPostContainerProps = IStateToProps & IDispatchToProps;

interface IPostContainerState {
  openEditDialog: boolean;
  openDeleteDialog: boolean;
}

class PostContainer extends Component<
  IPostContainerProps,
  IPostContainerState
> {
  constructor(props: IPostContainerProps) {
    super(props);
    this.state = {
      openEditDialog: false,
      openDeleteDialog: false,
    };
  }

  public componentDidMount() {
    const { fetchPosts, fetchUsers } = this.props;

    fetchPosts();
    fetchUsers();
  }
  public render() {
    const {
      posts,
      selectedPostId,
      loadingUsers,
      updatePost,
      deletePost,
    } = this.props;
    return (
      <ErrorBoundary>
        {selectedPostId &&
          !loadingUsers && (
            <Post
              {...this.props}
              post={posts[selectedPostId]}
              handleOpenEditDialog={this.handleOpenEditDialog}
              handleOpenDeleteDialog={this.handleOpenDeleteDialog}
            />
          )}
        {selectedPostId && (
          <PostEditDialog
            updatePost={updatePost}
            open={this.state.openEditDialog}
            handleCloseDialog={this.handleCloseEditDialog}
            post={posts[selectedPostId]}
          />
        )}
        {selectedPostId && (
          <PostDeleteDialog
            deletePost={deletePost}
            open={this.state.openDeleteDialog}
            handleCloseDialog={this.handleCloseDeleteDialog}
            post={posts[selectedPostId]}
          />
        )}
        <Route path={routes.selectedPost} component={CommentListContainer} />
        <Route path={routes.selectedPost} component={CommentAddContainer} />
      </ErrorBoundary>
    );
  }

  private handleOpenEditDialog = () => this.setState({ openEditDialog: true });
  private handleCloseEditDialog = () =>
    this.setState({ openEditDialog: false });
  private handleOpenDeleteDialog = () =>
    this.setState({ openDeleteDialog: true });
  private handleCloseDeleteDialog = () =>
    this.setState({ openDeleteDialog: false });
}

interface IStateToProps {
  posts: IPost[];
  users: IUser[];
  error?: string;
  loadingPosts: boolean;
  loadingUsers: boolean;
  selectedPostId?: number;
}

interface IDispatchToProps {
  fetchPosts: () => void;
  fetchUsers: () => void;
  deletePost: (id: number) => void;
  updatePost: (post: IPost) => void;
}

const mapStateToProps = (state: IStoreState) => ({
  error: postSelectors.getError(state),
  posts: postSelectors.getAllPostsObject(state),
  users: userSelectors.getAllUsersObject(state),
  loadingPosts: postSelectors.getLoadingStatus(state),
  loadingUsers: userSelectors.getLoadingStatus(state),
  selectedPostId: postSelectors.getSelectedPostId(state),
});

const mapDispatchToProps = (dispatch: Dispatch<IAction>) => ({
  fetchPosts: () => {
    dispatch(postActions.fetchPostsRequest());
  },
  updatePost: (post: IPost) => {
    dispatch(postActions.updatePostRequest(post));
  },
  deletePost: (id: number) => {
    dispatch(postActions.deletePostRequest(id));
    dispatch(push(routes.posts));
  },
  fetchUsers: () => {
    dispatch(userActions.fetchUsersRequest());
  },
});

export default connect<IStateToProps, IDispatchToProps, any>(
  mapStateToProps,
  mapDispatchToProps
)(PostContainer);
