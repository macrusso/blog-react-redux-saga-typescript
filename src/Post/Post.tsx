import React from 'react';
import { Item } from 'semantic-ui-react';
import { IPost } from '.';
import { IUser } from '../User';

export interface IPostProps {
  posts: IPost[];
  users: IUser[];
  loading: boolean;
}

const Post = ({
  users,
  posts,
  loading,
}: IPostProps) => (
    <Item.Group>
      {!loading && posts.length > 0 &&
        posts.map(post => (

          <Item key={post.id}>
            <Item.Image size='tiny' src='https://react.semantic-ui.com/images/wireframe/image.png' />

            <Item.Content>
              <Item.Header>{post.title}</Item.Header>
              <Item.Meta>by joe doe</Item.Meta>
              <Item.Description>
                {post.body}
              </Item.Description>
            </Item.Content>
          </Item>
        ))}
    </Item.Group>
  );

export default Post;