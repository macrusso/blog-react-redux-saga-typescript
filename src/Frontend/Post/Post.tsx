import React from 'react';
import { IPost, IUser } from '../../Entities';
import { Edit, Delete } from '@material-ui/icons';
import {
  Typography,
  withStyles,
  Chip,
  Avatar,
  IconButton,
} from '@material-ui/core';
import { NotFound } from '../Shared';

const styles = {
  card: {
    minWidth: 275,
  },
  bullet: {
    display: 'inline-block',
    margin: '0 2px',
    transform: 'scale(0.8)',
  },
  title: {
    fontSize: 14,
  },
  pos: {
    marginBottom: 12,
  },
  root: {
    minWidth: '100%',
    padding: 20,
    margin: '5px 0',
  },
};

interface IPostProps {
  post?: IPost;
  classes: any;
  users: IUser[];
  handleOpenEditDialog: () => void;
  handleOpenDeleteDialog: () => void;
}

const Post: React.SFC<IPostProps> = props => {
  const {
    post,
    classes,
    users,
    handleOpenEditDialog,
    handleOpenDeleteDialog,
  } = props;
  return (
    <>
      {post ? (
        <>
          <Typography variant="h5" component="h3">
            {post.title}
          </Typography>
          {post.userId && (
            <Chip
              avatar={
                <Avatar>{users[post.userId].name[0].toUpperCase()}</Avatar>
              }
              label={users[post.userId].name}
              className={classes.chip}
            />
          )}
          <Typography variant="body2">{post.body}</Typography>
          <IconButton
            aria-label="Delete"
            onClick={() => handleOpenDeleteDialog()}
          >
            <Delete />
          </IconButton>
          <IconButton aria-label="Edit" onClick={() => handleOpenEditDialog()}>
            <Edit />
          </IconButton>
        </>
      ) : (
        <NotFound />
      )}
    </>
  );
};

export default withStyles(styles)(Post);
