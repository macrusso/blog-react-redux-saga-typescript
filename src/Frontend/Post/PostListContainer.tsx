import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Dispatch } from 'redux';
import { IAction, IStoreState } from '../../types';
import ErrorBoundary from '../Common/ErrorBoundary';
import { postActions, postSelectors, IPost } from '../../Entities/Post';
import { userActions, userSelectors, IUser } from '../../Entities/User';
import { PostListItem } from '.';
import { PostAddContainer } from '../PostAdd';
import { Route } from 'react-router';
import { posts } from '../../routes';

type IPostListContainerProps = IStateToProps & IDispatchToProps;

class PostListContainer extends Component<IPostListContainerProps> {
  public componentDidMount() {
    const { fetchPosts, fetchUsers } = this.props;

    fetchPosts();
    fetchUsers();
  }

  public render() {
    const props = this.props;

    return (
      <ErrorBoundary>
        <PostAddContainer />
        <PostListItem {...props} />
      </ErrorBoundary>
    );
  }
}

interface IStateToProps {
  error?: string;
  loading: boolean;
  posts: IPost[];
  users: IUser[];
  usersLoading: boolean;
}

interface IDispatchToProps {
  fetchPosts: () => void;
  fetchUsers: () => void;
}

const mapStateToProps = (state: IStoreState) => ({
  error: postSelectors.getError(state),
  loading: postSelectors.getLoadingStatus(state),
  posts: postSelectors.getAllPosts(state),
  users: userSelectors.getAllUsersObject(state),
  usersLoading: userSelectors.getLoadingStatus(state),
});

const mapDispatchToProps = (dispatch: Dispatch<IAction>) => ({
  fetchPosts: () => {
    dispatch(postActions.fetchPostsRequest({}));
  },
  fetchUsers: () => {
    dispatch(userActions.fetchUsersRequest({}));
  },
});

export default connect<IStateToProps, IDispatchToProps, any>(
  mapStateToProps,
  mapDispatchToProps
)(PostListContainer);
