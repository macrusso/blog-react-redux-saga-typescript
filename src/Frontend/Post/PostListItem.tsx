import React from 'react';
import { Item, Container } from 'semantic-ui-react';
import { IPost } from '../../Entities/Post';
import { IUser } from '../../Entities/User';
import { Link } from 'react-router-dom';
import { selectedPost } from '../../routes';

export interface IPostListItemProps {
  posts: IPost[];
  users: IUser[];
  loading: boolean;
  usersLoading: boolean;
}

const PostListItem = ({
  users,
  posts,
  loading,
  usersLoading,
}: IPostListItemProps) => (
  <Container>
    <Item.Group>
      {!loading &&
        !usersLoading &&
        posts.length > 0 &&
        posts.map(post => (
          <Item key={post.id}>
            <Item.Image
              size="tiny"
              src="https://react.semantic-ui.com/images/wireframe/image.png"
            />
            <Item.Content>
              <Item.Header>
                <Link to={selectedPost.replace(':postId', post.id.toString())}>
                  {post.title}
                </Link>
              </Item.Header>
              <Item.Meta>by {users[post.userId].name}</Item.Meta>
              <Item.Description>{post.body}</Item.Description>
            </Item.Content>
          </Item>
        ))}
    </Item.Group>
  </Container>
);

export default PostListItem;
