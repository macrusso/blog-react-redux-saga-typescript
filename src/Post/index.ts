import * as postConstants from './postConstants';
import * as postSelectors from './postSelectors';
import * as postActions from './postActions';
import PostContainer from './PostContainer';
import * as postSagas from './postSagas';
import postReducer from './postReducer';
import IPost from './IPost';
import Post from './Post';

export {
  Post,
  IPost,
  postSagas,
  postActions,
  postReducer,
  postSelectors,
  postConstants,
  PostContainer,
};