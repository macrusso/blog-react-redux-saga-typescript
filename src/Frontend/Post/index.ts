import Post from './Post';
import PostContainer from './PostContainer';
import PostListItem from './PostListItem';
import PostListContainer from './PostListContainer';

export { Post, PostContainer, PostListItem, PostListContainer };
