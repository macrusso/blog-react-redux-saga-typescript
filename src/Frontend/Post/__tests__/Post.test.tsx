import React from 'react';
import { BrowserRouter } from 'react-router-dom';
import { shallow } from 'enzyme';
import Post from '../Post';
import { IPost, IUser } from '../../../Entities';

describe('Post', () => {
  test('Snapshot matches', () => {
    const post: IPost = {
      _id: '5bcc554259e0dbfda6ed1a55',
      userId: '5bcc554259e0dbfda6ed1a55',
      title: 'string',
      body: 'string',
    };

    const users: IUser[] = [
      {
        _id: '5bcc554259e0dbfda6ed1a55',
        name: 'string',
        email: 'string',
      },
    ];
    const wrapper = shallow(
      <BrowserRouter>
        <Post
          post={post}
          users={users}
          handleOpenEditDialog={jest.fn()}
          handleOpenDeleteDialog={jest.fn()}
        />
      </BrowserRouter>
    );
    expect(wrapper).toMatchSnapshot();
  });
});
