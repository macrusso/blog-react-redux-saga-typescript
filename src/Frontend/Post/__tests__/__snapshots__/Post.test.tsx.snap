// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Post Snapshot matches 1`] = `ShallowWrapper {}`;
